pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -375,6 +401,214 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
+
+        # Create a master sha256sum.txt including the sha256sum.txt files from all subarches
+        # This is the file we will sign -- trust is transitive to the subarches
+        for dir in os.listdir(release_mirror_dir):
+            if not os.path.isdir(f"{release_mirror_dir}/{dir}"):
+                continue
+            for root, dirs, files in os.walk(f"{release_mirror_dir}/{dir}"):
+                if "sha256sum.txt" not in files:
+                    continue
+                with open(f"{root}/sha256sum.txt", "rb") as f:
+                    shasum = hashlib.sha256(f.read()).hexdigest()
+                with open(f"{release_mirror_dir}/sha256sum.txt", 'a') as f:  # write shasum to sha256sum.txt
+                    f.write(f"{shasum}  {dir}/sha256sum.txt\n")
+        util.log_dir_tree(release_mirror_dir)
+
+        # Publish the clients to our S3 bucket.
+        await exectools.cmd_assert_async(f"aws s3 sync --no-progress --exact-timestamps {base_to_mirror_dir}/multi s3://art-srv-enterprise/pub/openshift-v4/multi", stdout=sys.stderr)
+
+    def create_symlink(self, path_to_dir, log_tree, log_shasum):
+        # External consumers want a link they can rely on.. e.g. .../latest/openshift-client-linux.tgz .
+        # So whatever we extract, remove the version specific info and make a symlink with that name.
+        current_path = os.getcwd() # /mnt/workspace/jenkins/working/build_promote-assembly
+        # path_to_dir is relative path artcd_working/to_mirror/openshift-v4/aarch64/clients/ocp/4.13.0-rc.6
+        for f in os.listdir(path_to_dir):
+            if f.endswith(('.tar.gz', '.bz', '.zip', '.tgz')):
+                # Is this already a link?
+                if os.path.islink(f"{path_to_dir}/{f}"):
+                    continue
+                # example file names:
+                #  - openshift-client-linux-4.3.0-0.nightly-2019-12-06-161135.tar.gz
+                #  - openshift-client-mac-4.3.0-0.nightly-2019-12-06-161135.tar.gz
+                #  - openshift-install-mac-4.3.0-0.nightly-2019-12-06-161135.tar.gz
+                #  - openshift-client-linux-4.1.9.tar.gz
+                #  - openshift-install-mac-4.3.0-0.nightly-s390x-2020-01-06-081137.tar.gz
+                #  ...
+                # So, match, and store in a group, any character up to the point we find -DIGIT. Ignore everything else
+                # until we match (and store in a group) one of the valid file extensions.
+                match = re.match(r'^([^-]+)((-[^0-9][^-]+)+)-[0-9].*(tar.gz|tgz|bz|zip)$', f)
+                if match:
+                    new_name = match.group(1) + match.group(2) + '.' + match.group(4)
+                    # Create a symlink like openshift-client-linux.tgz => openshift-client-linux-4.3.0-0.nightly-2019-12-06-161135.tar.gz
+                    os.symlink(f, new_name)
+                    shutil.move(new_name, f"{path_to_dir}/{new_name}")
```

Comment:
This doesn't seem to be needed.

Reply:
the symlink is created under curent work dir

Reply:
pls directly create the symlink in "{path_to_dir}/{new_name}" rather than create and move.